api/rest/client/methods.go

```diff
@@ -87,39 +87,35 @@ func (c *defaultClient) Unpin(ci cid.Cid) error {
 }
 
 // PinPath resolves given path into a cid and performs the pin operation.
-func (c *defaultClient) PinPath(pinPath string, replicationFactorMin, replicationFactorMax int, name string) (cid.Cid, error) {
-	var ci api.CidSerial
-
-	escName := url.QueryEscape(name)
+func (c *defaultClient) PinPath(opts api.PinOptionsWithPath) (api.Pin, error) {
+	var pin api.PinSerial
+	escName := url.QueryEscape(opts.Name)
 	err := c.do(
 		"POST",
-		fmt.Sprintf(
-			"/pins/%s?replication-min=%d&replication-max=%d&name=%s",
-			api.TrimSlacesAndSpaces(pinPath),
-			replicationFactorMin,
-			replicationFactorMax,
-			escName,
+		path.Clean(
+			fmt.Sprintf(
+				"/pins/%s?replication-min=%d&replication-max=%d&name=%s",
+				strings.TrimSuffix(opts.Path, "/"), // path.Clean() does not remove trailing slash. A valid ipfs path can contain a trailing slash.
```

Comment:
OK, the problem is not that Go doesn't do POST (that topic is from 2011), but that that 301 redirects are followed with a GET request (https://github.com/golang/go/blob/master/src/net/http/client.go). The new 308 redirects would be handled correctly but Gorilla does not offer a way to use them.

Anycase, this means that Gorilla redirects are not good enough and we should disable them because we cannot rely on them for anything non-GET/HEAD (this means setting `SkipClean` to  and `StrictSlash` to true). You're right that customizing CheckRedirects would be overkill.

In the light of this, we can path.Clean() and TrumSuffix in the client for convenience, but better if you make a function.